File: api/src/HgRunner.php
<?php

require_once('AsyncRunner.php');

class HgRunner {
	/**
	 * 
	 * @var string
	 */
	public $repoPath;

	/**
	 * 
	 * @var bool
	 */
	public $logState;

	const DEFAULT_HG = "/var/vcs/public";

	/**
	 * 
	 * @param string $repoPath
	 * @throws ValidationException
	 */
	function __construct($repoPath = DEFAULT_HG) {
		if (is_dir($repoPath)) {
			$this->repoPath = $repoPath;
		} else {
			throw new ValidationException("repo '$repoPath' doesn't exist!");
		}
		$logState = false;
	}

	private function logEvent($message) {
		$logFilename = $this->repoPath . "/hgRunner.log";
		$timestamp = date("YMd \TH.i");
		if (!$this->logState) {
			$this->logState = true;
			file_put_contents($logFilename, "\n$timestamp\n", FILE_APPEND | LOCK_EX);
		}
		file_put_contents($logFilename, "$message\n", FILE_APPEND | LOCK_EX);
	}

	/**
	 * 
	 * @param string $filepath
	 * @throws HgException
	 * @throws UnrelatedRepoException
	 * @throws Exception
	 * @return AsyncRunner
	 */
	function unbundle($filepath) {
		if (!is_file($filepath)) {
			throw new HgException("bundle file '$filepath' is not a file!");
		}
		chdir($this->repoPath); // NOTE: I tried with -R and it didn't work for me. CP 2012-06

		// run hg incoming to make sure this bundle is related to the repo
		$cmd = "hg incoming $filepath";
		$this->logEvent("cmd: $cmd");
		$asyncRunner = new AsyncRunner($filepath . ".incoming");
		$asyncRunner->run($cmd);
		$asyncRunner->synchronize();
		$output = $asyncRunner->getOutput();
		if (preg_match('/abort:.*unknown parent/', $output)) {
			throw new UnrelatedRepoException("Project is unrelated!  (unrelated bundle pushed to repo)");
		}
		if (preg_match('/parent:\s*-1:/', $output)) {
			throw new UnrelatedRepoException("Project is unrelated!  (unrelated bundle pushed to repo)");
		}
		if (preg_match('/abort:.*not a Mercurial bundle/', $output)) {
			throw new Exception("Project cannot be updated!  (corrupt bundle pushed to repo)");
		}

		$cmd = "hg unbundle $filepath";
		$this->logEvent("cmd: $cmd");
		$asyncRunner = new AsyncRunner($filepath);
		$asyncRunner->run($cmd);
		return $asyncRunner;
	}
	
	function assertIsRelatedRepo($bundleFilePath) {
		
	}

	/**
	 * 
	 * @param string $revision
	 * @return AsyncRunner
	 */
	function update($revision = "") {
		chdir($this->repoPath);
		$cmd = "hg update $revision";
		$this->logEvent("cmd: $cmd");
		$asyncRunner = new AsyncRunner($this->repoPath . "/hg_update");
		$asyncRunner->run($cmd);
		return $asyncRunner;
	}

	/**
	 * @param string $baseHashes[] expects the data to be just the "hash" and NOT the branch information preceded by a colon
	 * @param string $bundleFilePath
	 */
	function makeBundle($baseHashes, $bundleFilePath) {
		chdir($this->repoPath); // NOTE: I tried with -R and it didn't work for me. CP 2012-06
		if (count($baseHashes) == 1 && $baseHashes[0] == "0") {
			$cmd = "hg bundle --all $bundleFilePath";
		} else {
			$cmd = "hg bundle ";
			$baseHashes = (array)$baseHashes; //I can't figure out why this is sometimes not an array
			foreach($baseHashes as $hash)
			{
				$cmd .= "--base $hash ";
			}
			$cmd .= $bundleFilePath;
		}
		$asyncRunner = new AsyncRunner($bundleFilePath);
		$asyncRunner->run($cmd);
		return $asyncRunner;
	}

	/**
	 * helper function, mostly for tests
	 * @param string $baseHashes[]
	 * @param string $bundleFilePath
	 */
	function makeBundleAndWaitUntilFinished($baseHashes, $bundleFilePath) {
		$asyncRunner = $this->makeBundle($baseHashes, $bundleFilePath);
		$asyncRunner->synchronize();
		return $asyncRunner;
	}

	/**
	 *
	 * @return string a baseHash (without branch information)
	 */

	function getTip() {
		$revisionArray = $this->getRevisions(0, 1);
		return substr($revisionArray[0], 0, strpos($revisionArray[0], ':'));
	}

	function getBranchTips() {
		chdir($this->repoPath);
		exec('hg branches', $output, $returnval);
		$revisionArray = array();
		foreach($output as $branch) {
			if($branch == '') {
				$branchName = 'default';
			} else {
				$branchName = substr($branch, 0, strpos($branch, ' '));
			}
			//append the first revision for the branch to the array
			$revisionArray = array_merge($revisionArray, $this->getRevisionsInternal(0, 1, $branchName));
		}
		$revisions = array();
		foreach($revisionArray as $hashandbranch) {
			$revisions[] = substr($hashandbranch, 0, strpos($hashandbranch, ":"));
		}
		return $revisions;
	}

	//this method will return an array containing revision hash branch pairs e.g. 'fb7a8f23394d:default'
	function getRevisions($offset, $quantity) {
		return $this->getRevisionsInternal($offset, $quantity, NULL);
	}

	//this method will return an array containing revision hash branch pairs e.g. 'fb7a8f23394d:default'
	function getRevisionsInternal($offset, $quantity, $branch) {
		if ($quantity < 1) {
			throw new ValidationException("quantity parameter much be larger than 0");
		}
		chdir($this->repoPath);
		// I believe ':' is illegal in branch names, it is in tag, so we will use that to split the hash and branch
		$cmd = 'hg log';
		if(!is_null($branch)) {
			$cmd .= ' -b ' . $branch;
		}
		$cmd .= ' --template "{node|short}:{branches}\n"';
		exec($cmd, $output, $returnval);
		if (count($output) == 0) {
			exec('hg tip --template "{rev}:{branches}\n"', $output2, $returnval);

			if (count($output2) == 1 and strpos($output2[0], "-1") === 0) { // starts with -1
				//in the case of a tip result like '-1:default' we will return '0:default' to signal the empty repo
				$output2[0] = preg_replace('/^-1/', '0', $output2[0]);
				return $output2; //
			}
			throw new HgException("command '$cmd' failed!\n");
		}
		return array_slice($output, $offset, $quantity);
	}

	function isValidBase($hashes) {
		if (!is_array($hashes)) {
			$hashes = array($hashes);
		}
		if (count($hashes) == 1 && $hashes[0] == "0") { // special case indicating revision 0
			return true;
		}
		$foundHash = 0;
		$q = 200;
		$i = 0;
		while($foundHash < count($hashes)) {
			$revisions = $this->getRevisions($i, $q);
			if (count($revisions) == 0) {
				return false;
			}
			foreach($revisions as $hashandbranch) {
				$rev = substr($hashandbranch, 0, strpos($hashandbranch, ":"));
				if (array_search($rev, $hashes, false) !== false) {
					$foundHash++;
					if($foundHash >= count($hashes)) {
						break;
					}
				}
			}
			$i += $q;
		}
		return true;
	}

	// helper function used in tests
	function addAndCheckInFile($filePath) {
		chdir($this->repoPath);
		$cmd = "hg add $filePath";
		exec(escapeshellcmd($cmd) . " 2> /dev/null", $output, $returnval);
		if ($returnval != 0) {
			throw new HgException("command '$cmd' failed!\n");
		}
		$cmd = "hg commit -u 'system' -m 'added file $filePath'";
		exec(escapeshellcmd($cmd) . " 2> /dev/null", $output, $returnval);
		if ($returnval != 0) {
			throw new HgException("command '$cmd' failed!\n");
		}
	}
}

?>

File: csharp/src/HgResume.Api/HgRunner.cs
using System.Text.RegularExpressions;

namespace HgResume.Api;

/// <summary>
/// Mirrors api/src/HgRunner.php. All Mercurial operations shell out to the `hg` binary; stdout is
/// parsed verbatim (same templates/regexes as the PHP original) so behaviour matches byte-for-byte.
/// The PHP code chdir()'d into the repo before each call; we pass WorkingDirectory instead.
/// </summary>
public sealed class HgRunner
{
    private static readonly Regex UnknownParent = new("abort:.*unknown parent", RegexOptions.Compiled);
    private static readonly Regex ParentMinusOne = new("parent:\\s*-1:", RegexOptions.Compiled);
    private static readonly Regex NotAMercurialBundle = new("abort:.*not a Mercurial bundle", RegexOptions.Compiled);

    public string RepoPath { get; }

    public HgRunner(string repoPath)
    {
        if (!Directory.Exists(repoPath))
        {
            throw new ValidationException($"repo '{repoPath}' doesn't exist!");
        }
        RepoPath = repoPath;
    }

    // ---- push side: two-phase validate then unbundle --------------------------------------------

    public void StartValidating(string filepath)
    {
        if (!File.Exists(filepath))
        {
            throw new HgException($"bundle file '{filepath}' is not a file!");
        }
        // run hg incoming to make sure this bundle is related to the repo
        GetValidationRunner(filepath).Run(RepoPath, "hg", "incoming", filepath);
    }

    /// <returns>true if validation finished, otherwise false (still running)</returns>
    public bool FinishValidating(string filepath)
    {
        var asyncRunner = GetValidationRunner(filepath);
        if (asyncRunner.WaitForIsComplete())
        {
            string output = asyncRunner.GetOutput();
            if (UnknownParent.IsMatch(output))
            {
                throw new UnrelatedRepoException("Project is unrelated!  (unrelated bundle pushed to repo)");
            }
            if (ParentMinusOne.IsMatch(output))
            {
                throw new UnrelatedRepoException("Project is unrelated!  (unrelated bundle pushed to repo)");
            }
            if (NotAMercurialBundle.IsMatch(output))
            {
                throw new HgException("Project cannot be updated!  (corrupt bundle pushed to repo)");
            }
            return true;
        }
        return false;
    }

    private static AsyncRunner GetValidationRunner(string filepath) => new(filepath + ".incoming");

    public AsyncRunner Unbundle(string filepath)
    {
        if (!File.Exists(filepath))
        {
            throw new HgException($"bundle file '{filepath}' is not a file!");
        }
        var asyncRunner = new AsyncRunner(filepath);
        asyncRunner.Run(RepoPath, "hg", "unbundle", filepath);
        return asyncRunner;
    }

    public AsyncRunner Update(string revision = "")
    {
        var asyncRunner = new AsyncRunner(Path.Combine(RepoPath, "hg_update"));
        if (string.IsNullOrEmpty(revision))
        {
            asyncRunner.Run(RepoPath, "hg", "update");
        }
        else
        {
            asyncRunner.Run(RepoPath, "hg", "update", revision);
        }
        return asyncRunner;
    }

    // ---- pull side: bundle creation -------------------------------------------------------------

    public AsyncRunner MakeBundle(IReadOnlyList<string> baseHashes, string bundleFilePath)
    {
        var args = new List<string> { "bundle" };
        if (baseHashes.Count == 1 && baseHashes[0] == "0")
        {
            args.Add("--all");
            args.Add(bundleFilePath);
        }
        else
        {
            foreach (var hash in baseHashes)
            {
                args.Add("--base");
                args.Add(hash);
            }
            args.Add(bundleFilePath);
        }
        args.Add("-t");
        args.Add("v1");

        var asyncRunner = new AsyncRunner(bundleFilePath);
        asyncRunner.Run(RepoPath, "hg", args.ToArray());
        return asyncRunner;
    }

    public AsyncRunner MakeBundleAndWaitUntilFinished(IReadOnlyList<string> baseHashes, string bundleFilePath)
    {
        var asyncRunner = MakeBundle(baseHashes, bundleFilePath);
        if (!asyncRunner.WaitForIsComplete())
        {
            throw new HgException("Error: make bundle failed to complete");
        }
        return asyncRunner;
    }

    // ---- revision inspection --------------------------------------------------------------------

    /// <returns>a baseHash (without branch information)</returns>
    public string GetTip()
    {
        var revisionArray = GetRevisions(0, 1);
        string first = revisionArray[0];
        int colon = first.IndexOf(':');
        return colon >= 0 ? first.Substring(0, colon) : first;
    }

    public List<string> GetBranchTips()
    {
        var (branches, _) = ProcessRunner.RunSync(RepoPath, "hg", "branches");
        var revisionArray = new List<string>();
        foreach (var branch in branches)
        {
            string branchName;
            if (branch.Length == 0)
            {
                branchName = "default";
            }
            else
            {
                int space = branch.IndexOf(' ');
                branchName = space >= 0 ? branch.Substring(0, space) : branch;
            }
            revisionArray.AddRange(GetRevisionsInternal(0, 1, branchName));
        }
        var revisions = new List<string>();
        foreach (var hashAndBranch in revisionArray)
        {
            int colon = hashAndBranch.IndexOf(':');
            revisions.Add(colon >= 0 ? hashAndBranch.Substring(0, colon) : hashAndBranch);
        }
        return revisions;
    }

    /// <summary>Returns "hash:branch" pairs, e.g. 'fb7a8f23394d:default'.</summary>
    public List<string> GetRevisions(int offset, int quantity) => GetRevisionsInternal(offset, quantity, null);

    private List<string> GetRevisionsInternal(int offset, int quantity, string? branch)
    {
        if (quantity < 1)
        {
            throw new ValidationException("quantity parameter much be larger than 0");
        }
        // ':' is illegal in branch names (it is used in tags) so we use it to split hash and branch
        string[] args = branch is null
            ? new[] { "log", "--template", "{node|short}:{branches}\n" }
            : new[] { "log", "-b", branch, "--template", "{node|short}:{branches}\n" };

        var (output, _) = ProcessRunner.RunSync(RepoPath, "hg", args);
        if (output.Count == 0)
        {
            var (tip, _) = ProcessRunner.RunSync(RepoPath, "hg", "tip", "--template", "{rev}:{branches}\n");
            if (tip.Count == 1 && tip[0].StartsWith("-1"))
            {
                // Empty repo (hg init, zero changesets). At offset 0 we emit '0:<branch>' (from
                // '-1:<branch>') as the sentinel callers expect; past offset 0 there is nothing more,
                // so return empty. Returning the sentinel for every offset would make paginating
                // callers (e.g. IsValidBase) loop forever, since they never see an empty page.
                if (offset > 0)
                {
                    return new List<string>();
                }
                tip[0] = Regex.Replace(tip[0], "^-1", "0");
                return tip;
            }
            throw new HgException($"command 'hg log' failed!\n");
        }
        return output.Skip(offset).Take(quantity).ToList();
    }

    public bool IsValidBase(IReadOnlyList<string> hashes)
    {
        if (hashes.Count == 1 && hashes[0] == "0")
        {
            return true; // special case indicating revision 0
        }
        int foundHash = 0;
        const int q = 200;
        int i = 0;
        while (foundHash < hashes.Count)
        {
            var revisions = GetRevisions(i, q);
            if (revisions.Count == 0)
            {
                return false; // paged past the last revision without matching every hash
            }
            foreach (var hashAndBranch in revisions)
            {
                int colon = hashAndBranch.IndexOf(':');
                string rev = colon >= 0 ? hashAndBranch.Substring(0, colon) : hashAndBranch;
                if (hashes.Contains(rev))
                {
                    foundHash++;
                    if (foundHash >= hashes.Count) break;
                }
            }
            // A page shorter than the requested quantity means hg returned everything it had, so this
            // was the last page. Stop rather than advancing the offset again: this guarantees the loop
            // terminates even if GetRevisions ever returns a fixed non-empty page regardless of offset
            // (the empty-repo '0:' sentinel bug, or any similar future quirk).
            if (revisions.Count < q)
            {
                break;
            }
            i += q;
        }
        return foundHash >= hashes.Count;
    }
}
